modules/apps/dynamic-data-mapping/dynamic-data-mapping-api/src/main/java/com/liferay/dynamic/data/mapping/util/DDMDisplayRegistry.java

```diff
@@ -45,34 +45,35 @@ public String[] getPortletIds() {
 		return _getPortletIds();
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void setDDMDisplay(DDMDisplay ddmDisplay) {
-		_ddmDisplays.put(ddmDisplay.getPortletId(), ddmDisplay);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
+			bundleContext, DDMDisplay.class, null,
+			ServiceReferenceMapperFactory.create(
+				bundleContext,
+				(ddmDisplay, emitter) -> emitter.emit(
+					ddmDisplay.getPortletId())));
 	}
 
-	protected void unsetDDMDisplay(DDMDisplay ddmDisplay) {
-		_ddmDisplays.remove(ddmDisplay.getPortletId());
+	@Deactivate
+	protected void deactivate() {
+		_serviceTrackerMap.close();
 	}
 
 	private DDMDisplay _getDDMDisplay(String portletId) {
-		return _ddmDisplays.get(portletId);
+		return _serviceTrackerMap.getService(portletId);
```

Comment:
@W2jx Please add an SF commit to inline these private methods, as they seem to be only used by their public counterparts.

Reply:
I will fix it.